### PR TITLE
GH-38053: [C++][Go] Re-generate sources from Schema.fbs

### LINKED ISSUE #38053: [Go][C++] Re-generate flatbuffer sources
### Describe the enhancement requested

There are ~4 PRs depending on both string-view and list-view types that have been added to the `Schema.fbs`. Since generated files are checked-in, re-generating them once and for all will reduce noise in those PRs and remove the risk of conflicts (for instance, ordering of the types is significant as it affects the enum values assigned to each type).

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

Re-generate Go and C++ sources from Flatbuffers specs now including string-view and list-view types.

### What changes are included in this PR?

Documentation fixes on the .fbs files and the generated C++ and Go source files.

### Are these changes tested?

The files should be correct by construction. Existing tests guarantee that no mistake was made in re-generating the files.
* Closes: #38053

### CODE REVIEW

**Comment by bkietz** [comment]
Unfortunately, a string matching `/<a\b/` can't be used in comments in flatbuffer schemas (since the comments are copied to the generated c++, where they are recognized as ill formed html). I corrected this in my PR but if you're going to regenerate here then you'll need to modify Message.fbs

https://github.com/apache/arrow/pull/37792/files#diff-0623d567d0260222d5501b4e169141b5070eabc2ec09c3482da453a3346c5bf3R109

**Comment by felipecrv** [comment]
Thank you @bkietz. I pushed the fixes now and regenerated the sources again.

**Review by zeroshade — APPROVED** [review]
All looks good to me from a Go standpoint, i'll wait for someone to approve in addition before merging this

File: cpp/src/generated/Tensor_generated.h
Comment (pitrou) [inline cpp/src/generated/Tensor_generated.h:62]: These declarations don't seem to cause any problem, but they are weird nevertheless, as the corresponding definitions do not exist. Which version of Flatbuffers did you use?

Using flatc version 23.5.26 (the latest release), I don't get these declarations.

**Comment by pitrou** [comment]
Regardless of which flatc version is being used, it seems we should perhaps upgrade the [vendored flatbuffers headers](https://github.com/apache/arrow/tree/main/cpp/thirdparty/flatbuffers) to the same version.

For example, flatc 23.5.26 generates the following version check which seems to imply that the versions should match:
```c++
// Ensure the included flatbuffers.h is the same version as when this file was
// generated, otherwise it may not be compatible.
static_assert(FLATBUFFERS_VERSION_MAJOR == 23 &&
              FLATBUFFERS_VERSION_MINOR == 5 &&
              FLATBUFFERS_VERSION_REVISION == 26,
             "Non-compatible flatbuffers version included");
```

**Comment by pitrou** [comment]
@github-actions crossbow submit -g nightly-tests

**Review by bkietz — APPROVED** [review]

**Comment by pitrou** [comment]
> Regardless of which flatc version is being used, it seems we should perhaps upgrade the [vendored flatbuffers headers](https://github.com/apache/arrow/tree/main/cpp/thirdparty/flatbuffers) to the same version.

That said, as long as CI doesn't complain, we can merge and bump Flatbuffers later.

File: cpp/src/generated/Tensor_generated.h
Comment (felipecrv) [reply cpp/src/generated/Tensor_generated.h:62]: I used version `1.12.0` which is the version that generates the smallest diff and causes the least trouble.

**Comment by felipecrv** [comment]
> > Regardless of which flatc version is being used, it seems we should perhaps upgrade the [vendored flatbuffers headers](https://github.com/apache/arrow/tree/main/cpp/thirdparty/flatbuffers) to the same version.
> 
> That said, as long as CI doesn't complain, we can merge and bump Flatbuffers later.

The R-related failures seem unrelated.

**Comment by pitrou** [comment]
Agreed, I think we can merge now.

**Review by Fool256 — APPROVED** [review]

**Pull request merged**
